src/coreclr/gc/gcee.cpp

```diff
@@ -62,6 +62,11 @@ void GCHeap::ReportGenerationBounds()
     {
         g_theGCHeap->DiagDescrGenerations([](void*, int generation, uint8_t* rangeStart, uint8_t* rangeEnd, uint8_t* rangeEndReserved)
         {
+            // TODO: What shall we do with the GCGenerationRange event with frozen segments?
```

Comment:
The model that we are presenting to users is that these segments are not part of the GC so I would not expect them to be emitted in a GCGenerationRange event. I'd suggest we define a new event that triggers when the nongc segments are first allocated.

Reply:
> The more interesting part of this is that in order for the generation-aware analysis to work in PerfView, we need the ranges for the frozen segments

What is the dependency there? My understanding is that prior .NET 8 frozen segments were unused (aside from select special-case customers) so I wouldn't have expected any dependency to be there.

Reply:
> What is the dependency there? My understanding is that prior .NET 8 frozen segments were unused (aside from select special-case customers) so I wouldn't have expected any dependency to be there.

Generation Aware analysis depended on being able to determine the generation of every object it enumerates from the heap. Currently, it does so using the GCGenerationRange events.

It would be bad if heap enumeration comes up with an object that is not within any range. So either we have a range event for the frozen segments so that we can resolve its generation, or we make sure that the ETW-based heap enumeration skips frozen objects. In the latter case, PerfView will not be aware of the frozen object, and that may or may not be sufficient, I don't know for sure yet.

Reply:
> So either we have a range event for the frozen segments so that we can resolve its generation, or we make sure that the ETW-based heap enumeration skips frozen objects. In the latter case, PerfView will not be aware of the frozen object, and that may or may not be sufficient, I don't know for sure yet.

Given that the keyword is called `GCHeapCollectKeyword` and the model we are presenting to users is that objects in the nongc segments are not part of the GC heap then I'd say we shouldn't be enumerating nongc objects there. If needed in the future we could add a NonGCHeapCollectKeyword that has its own enumeration events.

Reply:
Overall it is a good idea. However, it is legal (and expected) to have objects in the GC heap referencing objects in the NonGC heap. Not sure how exactly we can split them cleanly.

Reply:
I think you could either pretend the references aren't present, assume references to out-of-bounds objects are immutable, or ask that we implement a NonGCHeap enumeration now.

Reply:
Actually, I prefer what I said in the initial message, leave it as is for now. We can address this set of work together (i.e. remove non-gc objects from GC heap enumeration, remove non-gc object bounds, add non-gc object enumeration, add non-gc bounds) together as a whole.

This will keep the analysis tooling working while we are still working on it.

Reply:
> leave it as is for now.

I just want to make sure we agree on how long 'for now' is :) Not touching it in this PR is fine, but by the time we ship .NET 8, non-gc segments should not be present in GCGenerationRange events.

Reply:
Sounds good to me.

Reply:
Fixed